File: hash_table/3sum.cpp
#include<bits/stdc++.h>
using namespace std;

    vector<vector<int>> threeSum(vector<int>& v) {
        
        vector<vector<int> > ans;
        sort(v.begin(),v.end());
         for(int i=0;i<v.size();i++)
         {
             int k=-v[i];
             int start=i+1;
             int end = v.size()-1;
             int sum=0;
             vector<int> temp;
             while(start<end)
             {
                 sum=v[start]+v[end];
                 if(sum>k)
                 {
                     end--;
                 }
                 else if(sum<k)
                 {
                     start++;
                 }
                 else
                 {
                     temp.push_back(-k);
                     temp.push_back(v[start]);
                     temp.push_back(v[end]);
                     break;
                 }
             }
             if(temp.size()>0)
             {
             	  ans.push_back(temp);
             }
           
         }
        return ans;
    }
int main()
{
	int t;
	cin>>t;
	while(t--)
	{
		 vector<vector<int> > ans;
		int n;
		cin>>n;
		vector<int> a(n);

		for(int i=0;i<n;i++)
		{
			cin>>a[i];
		}
		ans=threeSum(a);
		for(int i=0;i<ans.size();i++)
		{
			for(int j=0;j<ans[i].size();j++)
			{
				cout<<ans[i][j]<<" ";
			}
			cout<<endl;
		}
	}
}

File: hash_table/3sum.py
def _3sum(a):
	a=sorted(a)
	n=len(a)
	triplets = set()
	for i in range(n):
		k = -a[i]
		start=i+1
		end=n-1
		
		while start<end:
			s=a[start]+a[end]
			if s==k:
				triplets.add( (a[i],a[start],a[end]) )
				start+=1
				end-=1
			elif s>k:
				end-=1
			else:
				start+=1

	m = []
	for z in triplets:
		m.append( [z[0],z[1],z[2]] )
	return m


t = int( input() )
while t>0:
	n = int( input() )
	a = list( map(int,input().split()) )
	# print(a)
	ans = _3sum(a)
	print(ans)
	t-=1


